File: python/statgen/_bfile_utils.py
import re
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
from pandas.errors import EmptyDataError, ParserError, ParserWarning

from ._utils import CANONICAL_CHR_ORDER, CHR_RANK, allele_hash64
from ._variant_match import check_unique_sorted_variant_keys, numeric_variant_keys

IGNORED_CHR = {"Y", "MT"}
FAM_PUBLIC_COLUMNS = ["fid", "iid", "father_id", "mother_id", "sex"]

_BIM_NAMES = ["chr", "snp", "cm", "bp", "a1", "a2"]
_FAM_NAMES = ["fid", "iid", "father_id", "mother_id", "sex", "pheno"]
_PLOIDY_NAMES = ["ploidy_male", "ploidy_female"]
_DNA_ALLELE_RE = re.compile(r"^[ACGT]+$")
_BED_LOOKUP_INT8 = None
_BED_MAX_READ_BYTES = 64 * 1024 * 1024


def _read_csv_strict(path: Path, *, sep, names, description: str | None = None) -> pd.DataFrame:
    expected = description or f"{len(names)} delimited columns"
    try:
        with warnings.catch_warnings():
            warnings.simplefilter("error", ParserWarning)
            df = pd.read_csv(
                path,
                sep=sep,
                header=None,
                dtype=str,
                index_col=False,
                keep_default_na=False,
                na_filter=False,
            )
    except (EmptyDataError, ParserError, ParserWarning, ValueError) as exc:
        raise ValueError(f"{path}: expected {expected}") from exc
    if df.shape[1] != len(names):
        raise ValueError(f"{path}: expected {expected}, got {df.shape[1]}")
    df.columns = names
    return df


def _validate_source_sort_order(source_df: pd.DataFrame, path: Path) -> None:
    if source_df.empty:
        return
    line = (
        source_df["line"].to_numpy(dtype=np.int64)
        if "line" in source_df.columns
        else source_df.index.to_numpy(dtype=np.int64) + 1
    )
    chr_rank = source_df["chr"].map(CHR_RANK)
    bad_chr = chr_rank.isna()
    if bad_chr.any():
        idx = int(bad_chr.idxmax())
        raise ValueError(f"{path}:{idx + 1}: chr must use canonical labels 1-22 or X")

    keys = numeric_variant_keys(
        source_df["bp"].to_numpy(dtype=np.int64),
        allele_hash64(source_df["a1"].to_numpy(dtype=object)),
        allele_hash64(source_df["a2"].to_numpy(dtype=object)),
    )
    order_df = pd.DataFrame(
        {
            "chr_rank": chr_rank.to_numpy(dtype=np.int64),
            "bp": source_df["bp"].to_numpy(dtype=np.int64),
            "line": line,
        }
    )

    for rank in np.unique(order_df["chr_rank"].to_numpy(dtype=np.int64)):
        mask = order_df["chr_rank"].to_numpy(dtype=np.int64) == rank
        shard_keys = keys[mask]
        sorted_keys = np.sort(shard_keys, order=("bp", "a1_hash64", "a2_hash64"))
        check_unique_sorted_variant_keys(sorted_keys, f"BIM {path}", "variant")

    prev = order_df.shift(1)
    bad_order = (
        (order_df["chr_rank"] < prev["chr_rank"])
        | (
            (order_df["chr_rank"] == prev["chr_rank"])
            & (order_df["bp"] < prev["bp"])
        )
    ).fillna(False)
    if bad_order.any():
        line = int(order_df.loc[bad_order.idxmax(), "line"])
        raise ValueError(
            f"{path}:{line}: rows must be sorted by (chr_rank, bp) in canonical contig order"
        )


def _parse_bim(path: Path) -> pd.DataFrame:
    path = Path(path)
    df = _read_csv_strict(
        path,
        sep=r"\s+",
        names=_BIM_NAMES,
        description="6 whitespace-delimited columns",
    )
    df = df.copy()

    chr_col = df["chr"]
    bad_chr = chr_col.eq("")
    if bad_chr.any():
        idx = int(bad_chr.idxmax())
        raise ValueError(f"{path}:{idx + 1}: chr must be non-empty")

    chr_style = chr_col.str.lower().str.startswith("chr")
    if chr_style.any():
        idx = int(chr_style.idxmax())
        raise ValueError(f"{path}:{idx + 1}: chr-style labels (e.g., chr1/chrX) are not allowed")

    known_chr = chr_col.isin(CANONICAL_CHR_ORDER) | chr_col.isin(IGNORED_CHR)
    if not known_chr.all():
        idx = int((~known_chr).idxmax())
        raise ValueError(
            f"{path}:{idx + 1}: unsupported chr label {chr_col.iat[idx]!r}; expected 1-22, X (Y/MT are ignored)"
        )

    bad_allele = df["a1"].eq("") | df["a2"].eq("")
    if bad_allele.any():
        idx = int(bad_allele.idxmax())
        raise ValueError(f"{path}:{idx + 1}: a1 and a2 must be non-empty")

    bad_a1_syntax = ~df["a1"].str.fullmatch(_DNA_ALLELE_RE.pattern)
    if bad_a1_syntax.any():
        idx = int(bad_a1_syntax.idxmax())
        raise ValueError(
            f"{path}:{idx + 1}: a1 must be uppercase DNA bases (A/C/G/T): {df['a1'].iat[idx]!r}"
        )
    bad_a2_syntax = ~df["a2"].str.fullmatch(_DNA_ALLELE_RE.pattern)
    if bad_a2_syntax.any():
        idx = int(bad_a2_syntax.idxmax())
        raise ValueError(
            f"{path}:{idx + 1}: a2 must be uppercase DNA bases (A/C/G/T): {df['a2'].iat[idx]!r}"
        )
    same_allele = df["a1"].eq(df["a2"])
    if same_allele.any():
        idx = int(same_allele.idxmax())
        raise ValueError(f"{path}:{idx + 1}: a1 and a2 must differ")

    cm_num = pd.to_numeric(df["cm"], errors="coerce")
    bad_cm = cm_num.isna()
    if bad_cm.any():
        idx = int(bad_cm.idxmax())
        raise ValueError(f"{path}:{idx + 1}: cm is not a number: {df['cm'].iat[idx]!r}")

    bp_num = pd.to_numeric(df["bp"], errors="coerce")
    bad_bp = bp_num.isna() | (np.floor(bp_num) != bp_num)
    if bad_bp.any():
        idx = int(bad_bp.idxmax())
        raise ValueError(f"{path}:{idx + 1}: bp is not an integer: {df['bp'].iat[idx]!r}")
    df["bp"] = bp_num.astype(np.int64)

    _validate_source_sort_order(df.loc[df["chr"].isin(CANONICAL_CHR_ORDER)], path)
    return df.reset_index(drop=True)


def _parse_fam(path: Path) -> pd.DataFrame:
    path = Path(path)
    df = _read_csv_strict(
        path,
        sep=r"\s+",
        names=_FAM_NAMES,
        description="6 whitespace-delimited columns",
    )
    required = df[_FAM_NAMES].eq("")
    if required.any().any():
        row = int(np.flatnonzero(required.any(axis=1).to_numpy())[0] + 1)
        raise ValueError(f"{path}:{row}: FAM must contain exactly 6 non-empty columns")

    sex_num = pd.to_numeric(df["sex"], errors="coerce")
    bad_sex = sex_num.isna() | (np.floor(sex_num) != sex_num) | ~sex_num.isin([0, 1, 2])
    if bad_sex.any():
        idx = int(bad_sex.idxmax())
        raise ValueError(f"{path}:{idx + 1}: FAM sex must be one of PLINK values 1, 2, or 0")

    out = df[["fid", "iid", "father_id", "mother_id"]].copy()
    out["sex"] = sex_num.astype(np.int8)
    dup = out.duplicated(subset=["fid", "iid"], keep="first")
    if dup.any():
        idx = int(dup.idxmax())
        raise ValueError(f"{path}:{idx + 1}: duplicate FAM subject pair (fid, iid)")
    return out.reset_index(drop=True)


def _parse_ploidy(path: Path | None, source_num_snp: int, source_chr=None) -> tuple[np.ndarray, np.ndarray]:
    if path is None:
        ploidy_male = np.full(source_num_snp, 2.0, dtype=float)
        ploidy_female = np.full(source_num_snp, 2.0, dtype=float)
        if source_chr is not None:
            chrs = np.asarray(source_chr, dtype=object).reshape(-1)
            if chrs.size != source_num_snp:
                raise ValueError("source_chr length must match source_num_snp")
            ploidy_male[chrs == "X"] = 1.0
        return ploidy_male, ploidy_female

    path = Path(path)
    df = _read_csv_strict(
        path,
        sep="\t",
        names=_PLOIDY_NAMES,
        description="2 tab-delimited columns",
    )
    if df.shape[0] != source_num_snp:
        raise ValueError(
            f"{path}: PLOIDY row count mismatch: expected {source_num_snp}, got {df.shape[0]}"
        )
    numeric = df[_PLOIDY_NAMES].apply(pd.to_numeric, errors="coerce")
    bad_numeric = numeric.isna()
    if bad_numeric.any().any():
        row = int(np.flatnonzero(bad_numeric.any(axis=1).to_numpy())[0] + 1)
        raise ValueError(f"{path}:{row}: ploidy values must be integers 0, 1, or 2")
    values = numeric.to_numpy(dtype=np.int64)
    bad = ~np.isin(values, [0, 1, 2])
    if bad.any():
        row = int(np.flatnonzero(bad.any(axis=1))[0] + 1)
        raise ValueError(f"{path}:{row}: ploidy values must be 0, 1, or 2")
    return (
        numeric["ploidy_male"].to_numpy(dtype=float),
        numeric["ploidy_female"].to_numpy(dtype=float),
    )


def _expected_bed_size(source_num_sample: int, source_num_snp: int) -> int:
    return 3 + ((int(source_num_sample) + 3) // 4) * int(source_num_snp)


def _validate_bed(path: Path, source_num_sample: int, source_num_snp: int) -> int:
    path = Path(path)
    with open(path, "rb") as f:
        magic = f.read(3)
    if len(magic) != 3 or magic[:2] != b"\x6c\x1b":
        raise ValueError(f"{path}: invalid PLINK BED magic bytes")
    if magic[2] != 1:
        raise ValueError(f"{path}: PLINK BED must be SNP-major mode")
    size = path.stat().st_size
    expected = _expected_bed_size(source_num_sample, source_num_snp)
    if size != expected:
        raise ValueError(f"{path}: BED file size mismatch: expected {expected}, got {size}")
    return int(size)


def _bed_lookup_int8() -> np.ndarray:
    global _BED_LOOKUP_INT8
    if _BED_LOOKUP_INT8 is None:
        values = np.arange(256, dtype=np.uint16)[:, None]
        shifts = (np.arange(4, dtype=np.uint16) * 2)[None, :]
        two_bit = ((values >> shifts) & 0b11).astype(np.uint8)
        lookup = np.empty((256, 4), dtype=np.int8)
        lookup[two_bit == 0b00] = 2
        lookup[two_bit == 0b01] = -1
        lookup[two_bit == 0b10] = 1
        lookup[two_bit == 0b11] = 0
        lookup.setflags(write=False)
        _BED_LOOKUP_INT8 = lookup
    return _BED_LOOKUP_INT8


def _read_bed_rows_int8(path: Path, source_rows: np.ndarray, source_num_sample: int) -> np.ndarray:
    path = Path(path)
    source_rows = np.asarray(source_rows, dtype=np.int64).reshape(-1)
    n_samples = int(source_num_sample)
    bytes_per_snp = (n_samples + 3) // 4
    if source_rows.size == 0:
        return np.empty((n_samples, 0), dtype=np.int8)
    unique_rows, inverse = np.unique(source_rows, return_inverse=True)
    unique_decoded = np.empty((n_samples, unique_rows.size), dtype=np.int8)
    rows_per_read = unique_rows.size if bytes_per_snp == 0 else max(1, _BED_MAX_READ_BYTES // bytes_per_snp)
    lookup = _bed_lookup_int8()
    with open(path, "rb") as f:
        group_starts = np.r_[0, np.flatnonzero(np.diff(unique_rows) != 1) + 1]
        group_stops = np.r_[group_starts[1:], unique_rows.size]
        # PERF: one seek/read per contiguous BED row block; requested order and
        #       repeats are restored after block decoding. Blocks are capped to
        #       bound transient packed/decode allocations for large requests.
        for start, stop in zip(group_starts, group_stops, strict=True):
            for chunk_start in range(start, stop, rows_per_read):
                chunk_stop = min(chunk_start + rows_per_read, stop)
                block_rows = unique_rows[chunk_start:chunk_stop]
                expected_bytes = int(bytes_per_snp * block_rows.size)
                f.seek(3 + int(block_rows[0]) * bytes_per_snp)
                packed = f.read(expected_bytes)
                if len(packed) != expected_bytes:
                    raise ValueError(
                        f"{path}: short BED read for source SNP rows {int(block_rows[0])}-{int(block_rows[-1])}: "
                        f"expected {expected_bytes} bytes, got {len(packed)}"
                    )
                byte_values = np.frombuffer(packed, dtype=np.uint8)
                block_decoded = lookup[byte_values].reshape(block_rows.size, bytes_per_snp * 4)
                unique_decoded[:, chunk_start:chunk_stop] = block_decoded[:, :n_samples].T
    return unique_decoded[:, inverse]

File: R-package/R/bfile_utils.R
.parse_bim <- function(path) {
  if (!file.exists(path)) {
    stop(sprintf("File not found: %s", path), call. = FALSE)
  }
  df <- tryCatch(
    utils::read.table(
      path,
      header = FALSE,
      sep = "",
      quote = "",
      comment.char = "",
      stringsAsFactors = FALSE,
      colClasses = rep("character", 6L),
      fill = FALSE
    ),
    error = function(e) {
      stop(sprintf("%s: expected 6 whitespace-delimited columns", path), call. = FALSE)
    }
  )
  if (ncol(df) != 6L) {
    stop(sprintf("%s: expected 6 whitespace-delimited columns, got %d", path, ncol(df)), call. = FALSE)
  }
  names(df) <- c("chr", "snp", "cm", "bp", "a1", "a2")
  bad_chr <- is.na(df$chr) | df$chr == ""
  if (any(bad_chr)) {
    stop(sprintf("%s:%d: chr must be non-empty", path, which(bad_chr)[[1]]), call. = FALSE)
  }
  chr_style <- startsWith(tolower(df$chr), "chr")
  if (any(chr_style)) {
    stop(sprintf("%s:%d: chr-style labels (e.g., chr1/chrX) are not allowed", path, which(chr_style)[[1]]), call. = FALSE)
  }
  known_chr <- df$chr %in% c(.canonical_chr_order, .ignored_chr)
  if (!all(known_chr)) {
    idx <- which(!known_chr)[[1]]
    stop(sprintf("%s:%d: unsupported chr label %s; expected 1-22, X (Y/MT are ignored)", path, idx, sQuote(df$chr[[idx]])), call. = FALSE)
  }
  bad_allele <- is.na(df$a1) | is.na(df$a2) | df$a1 == "" | df$a2 == ""
  if (any(bad_allele)) {
    stop(sprintf("%s:%d: a1 and a2 must be non-empty", path, which(bad_allele)[[1]]), call. = FALSE)
  }
  bad_a1 <- !grepl("^[ACGT]+$", df$a1)
  if (any(bad_a1)) {
    idx <- which(bad_a1)[[1]]
    stop(sprintf("%s:%d: a1 must be uppercase DNA bases (A/C/G/T): %s", path, idx, sQuote(df$a1[[idx]])), call. = FALSE)
  }
  bad_a2 <- !grepl("^[ACGT]+$", df$a2)
  if (any(bad_a2)) {
    idx <- which(bad_a2)[[1]]
    stop(sprintf("%s:%d: a2 must be uppercase DNA bases (A/C/G/T): %s", path, idx, sQuote(df$a2[[idx]])), call. = FALSE)
  }
  cm <- suppressWarnings(as.numeric(df$cm))
  if (any(is.na(cm) | !is.finite(cm))) {
    idx <- which(is.na(cm) | !is.finite(cm))[[1]]
    stop(sprintf("%s:%d: cm is not a number: %s", path, idx, sQuote(df$cm[[idx]])), call. = FALSE)
  }
  bp_num <- suppressWarnings(as.numeric(df$bp))
  bad_bp <- is.na(bp_num) | !is.finite(bp_num) | floor(bp_num) != bp_num
  if (any(bad_bp)) {
    idx <- which(bad_bp)[[1]]
    stop(sprintf("%s:%d: bp is not an integer: %s", path, idx, sQuote(df$bp[[idx]])), call. = FALSE)
  }
  df$bp <- as.integer(bp_num)
  df$a1_hash64 <- .allele_hash64(df$a1)
  df$a2_hash64 <- .allele_hash64(df$a2)
  .validate_bim_order_and_duplicates(df[df$chr %in% .canonical_chr_order, , drop = FALSE], path)
  df
}

.validate_bim_order_and_duplicates <- function(df, path) {
  if (!nrow(df)) {
    return(invisible(NULL))
  }
  rank <- unname(.chr_rank[df$chr])
  bad_order <- c(FALSE, rank[-1L] < rank[-length(rank)] | (rank[-1L] == rank[-length(rank)] & df$bp[-1L] < df$bp[-nrow(df)]))
  if (any(bad_order)) {
    stop(sprintf("%s:%d: rows must be sorted by (chr_rank, bp) in canonical contig order", path, which(bad_order)[[1]]), call. = FALSE)
  }

  for (label in unique(df$chr)) {
    keep <- df$chr == label
    .check_unique_variant_keys(
      df$bp[keep],
      df$a1_hash64[keep],
      df$a2_hash64[keep],
      sprintf("%s: shard %s", path, label),
      "variant"
    )
  }
  invisible(NULL)
}

.parse_fam <- function(path) {
  df <- tryCatch(
    utils::read.table(
      path,
      header = FALSE,
      sep = "",
      quote = "",
      comment.char = "",
      stringsAsFactors = FALSE,
      colClasses = rep("character", 6L),
      fill = FALSE
    ),
    error = function(e) {
      stop(sprintf("%s: expected 6 whitespace-delimited columns", path), call. = FALSE)
    }
  )
  if (ncol(df) != 6L) {
    stop(sprintf("%s: expected 6 whitespace-delimited columns, got %d", path, ncol(df)), call. = FALSE)
  }
  names(df) <- c("fid", "iid", "father_id", "mother_id", "sex", "pheno")
  bad_required <- df == ""
  if (any(bad_required)) {
    stop(sprintf("%s:%d: FAM fields must be non-empty; 6 whitespace-delimited columns are required", path, which(rowSums(bad_required) > 0L)[[1]]), call. = FALSE)
  }
  sex_num <- suppressWarnings(as.numeric(df$sex))
  bad_sex <- is.na(sex_num) | !is.finite(sex_num) | floor(sex_num) != sex_num | !(sex_num %in% c(0, 1, 2))
  if (any(bad_sex)) {
    stop(sprintf("%s:%d: FAM sex must be one of PLINK values 1, 2, or 0", path, which(bad_sex)[[1]]), call. = FALSE)
  }
  out <- df[c("fid", "iid", "father_id", "mother_id")]
  out$sex <- as.integer(sex_num)
  dup <- duplicated(paste(out$fid, out$iid, sep = "\r"))
  if (any(dup)) {
    stop(sprintf("%s:%d: duplicate FAM subject pair (fid, iid)", path, which(dup)[[1]]), call. = FALSE)
  }
  out
}

.parse_ploidy <- function(path, source_num_snp, source_chr = NULL) {
  if (is.null(path)) {
    male <- rep.int(2, source_num_snp)
    female <- rep.int(2, source_num_snp)
    if (!is.null(source_chr)) {
      if (length(source_chr) != source_num_snp) {
        stop("source_chr length must match source_num_snp", call. = FALSE)
      }
      male[as.character(source_chr) == "X"] <- 1
    }
    return(list(male = as.numeric(male), female = as.numeric(female)))
  }
  df <- tryCatch(
    utils::read.table(
      path,
      header = FALSE,
      sep = "\t",
      quote = "",
      comment.char = "",
      stringsAsFactors = FALSE,
      colClasses = rep("character", 2L),
      fill = FALSE
    ),
    error = function(e) {
      stop(sprintf("%s: expected 2 tab-delimited columns", path), call. = FALSE)
    }
  )
  if (ncol(df) != 2L) {
    stop(sprintf("%s: expected 2 tab-delimited columns, got %d", path, ncol(df)), call. = FALSE)
  }
  if (nrow(df) != source_num_snp) {
    stop(sprintf("%s: PLOIDY row count mismatch: expected %d, got %d", path, source_num_snp, nrow(df)), call. = FALSE)
  }
  values <- matrix(
    c(
      suppressWarnings(as.numeric(df[[1L]])),
      suppressWarnings(as.numeric(df[[2L]]))
    ),
    ncol = 2L
  )
  bad <- is.na(values) | !is.finite(values) | floor(values) != values | !(values %in% c(0, 1, 2))
  if (any(bad)) {
    stop(sprintf("%s:%d: ploidy values must be integers 0, 1, or 2", path, which(rowSums(bad) > 0L)[[1]]), call. = FALSE)
  }
  list(male = as.numeric(values[, 1]), female = as.numeric(values[, 2]))
}

.expected_bed_size <- function(source_num_sample, source_num_snp) {
  sample_count <- as.numeric(source_num_sample)
  snp_count <- as.numeric(source_num_snp)
  3 + floor((sample_count + 3) / 4) * snp_count
}

.validate_bed_file <- function(path, source_num_sample, source_num_snp) {
  if (!file.exists(path)) {
    stop(sprintf("File not found: %s", path), call. = FALSE)
  }
  con <- file(path, "rb")
  on.exit(close(con), add = TRUE)
  magic <- readBin(con, what = "raw", n = 3L)
  if (length(magic) != 3L || !identical(magic[1:2], as.raw(c(0x6c, 0x1b)))) {
    stop(sprintf("%s: invalid PLINK BED magic bytes", path), call. = FALSE)
  }
  if (!identical(magic[[3]], as.raw(0x01))) {
    stop(sprintf("%s: PLINK BED must be SNP-major mode", path), call. = FALSE)
  }
  size <- file.info(path)$size
  expected <- .expected_bed_size(source_num_sample, source_num_snp)
  if (!identical(as.numeric(size), as.numeric(expected))) {
    stop(sprintf("%s: BED file size mismatch: expected %.0f, got %.0f", path, expected, size), call. = FALSE)
  }
  as.numeric(size)
}

.bed_lookup_int <- local({
  values <- 0:255
  out <- matrix(0L, nrow = 256L, ncol = 4L)
  for (j in 0:3) {
    two_bit <- bitwAnd(bitwShiftR(values, 2L * j), 3L)
    decoded <- integer(256L)
    decoded[two_bit == 0L] <- 2L
    decoded[two_bit == 1L] <- -1L
    decoded[two_bit == 2L] <- 1L
    decoded[two_bit == 3L] <- 0L
    out[, j + 1L] <- decoded
  }
  out
})

.bed_max_read_bytes <- 64 * 1024^2

.read_bed_rows_int8_codes <- function(path, source_rows0, source_num_sample) {
  source_rows0 <- as.integer(source_rows0)
  sample_count <- as.numeric(source_num_sample)
  bytes_per_snp <- floor((sample_count + 3) / 4)
  if (!length(source_rows0)) {
    return(matrix(0L, nrow = source_num_sample, ncol = 0L))
  }
  unique_rows0 <- sort(unique(source_rows0))
  inverse <- match(source_rows0, unique_rows0)
  unique_decoded <- matrix(0L, nrow = source_num_sample, ncol = length(unique_rows0))
  rows_per_read <- if (bytes_per_snp == 0) {
    length(unique_rows0)
  } else {
    max(1L, as.integer(floor(.bed_max_read_bytes / bytes_per_snp)))
  }
  con <- file(path, "rb")
  on.exit(close(con), add = TRUE)

  breaks <- c(1L, which(diff(unique_rows0) != 1L) + 1L, length(unique_rows0) + 1L)
  for (g in seq_len(length(breaks) - 1L)) {
    group_cols <- seq.int(breaks[[g]], breaks[[g + 1L]] - 1L)
    chunk_starts <- seq.int(1L, length(group_cols), by = rows_per_read)
    for (chunk_start in chunk_starts) {
      chunk_stop <- min(chunk_start + rows_per_read - 1L, length(group_cols))
      block_cols <- group_cols[chunk_start:chunk_stop]
      block_rows0 <- unique_rows0[block_cols]
      byte_offset <- 3 + as.numeric(block_rows0[[1]]) * bytes_per_snp
      expected_bytes <- bytes_per_snp * length(block_rows0)
      if (expected_bytes > .Machine$integer.max) {
        stop(sprintf("%s: BED read block is too large: %.0f bytes", path, expected_bytes), call. = FALSE)
      }
      seek(con, where = byte_offset, origin = "start", rw = "read")
      packed <- readBin(con, what = "integer", n = as.integer(expected_bytes), size = 1L, signed = FALSE, endian = "little")
      if (length(packed) != expected_bytes) {
        stop(
          sprintf("%s: short BED read for source SNP rows %d-%.0f: expected %.0f bytes, got %d", path, block_rows0[[1]], block_rows0[[length(block_rows0)]], expected_bytes, length(packed)),
          call. = FALSE
        )
      }
      decoded_flat <- as.vector(t(.bed_lookup_int[packed + 1L, , drop = FALSE]))
      decoded_block <- matrix(decoded_flat, nrow = bytes_per_snp * 4, ncol = length(block_rows0))
      unique_decoded[, block_cols] <- decoded_block[seq_len(sample_count), , drop = FALSE]
    }
  }
  unique_decoded[, inverse, drop = FALSE]
}
